### ISSUE TITLE
should_implement_trait shouldn't emit for from_str methods on types with lifetimes

### ISSUE THREAD

**Issue by DianaNites** [rust-lang/rust-clippy#9762]
### Summary

Clippy wrongly suggests implementing an incompatible trait

### Lint Name

should_implement_trait

### Reproducer

I tried this code:

```rust
pub struct HoldsStr<'a> {
    _s: &'a str,
}

impl<'a> HoldsStr<'a> {
    pub fn from_str(_s: &'a str) -> Self {
        Self { _s }
    }
}
```

I saw this happen:

```
warning: method `from_str` can be confused for the standard trait method `std::str::FromStr::from_str`
 --> src/lib.rs:6:5
  |
6 | /     pub fn from_str(_s: &'a str) -> Self {
7 | |         Self { _s }
8 | |     }
  | |_____^
  |
  = help: consider implementing the trait `std::str::FromStr` or choosing a less ambiguous method name
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#should_implement_trait
  = note: `#[warn(clippy::should_implement_trait)]` on by default
```

I expected to see this happen:

No warnings

### Version

```text
rustc 1.64.0 (a55dd71d5 2022-09-19)
binary: rustc
commit-hash: a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52
commit-date: 2022-09-19
host: x86_64-unknown-linux-gnu
release: 1.64.0
LLVM version: 14.0.6
```

**Comment by pheki** [comment]
Also this lint is triggered in the exact same way for the official serde deserializer example: https://serde.rs/impl-deserializer.html
